### PR TITLE
Merge branch release/3.1 --> master

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Better PR for https://github.com/dotnet/aspnetcore/pull/18896

### CODE REVIEW

File: src/Razor/Razor.Runtime/ref/Microsoft.AspNetCore.Razor.Runtime.Manual.cs [hunk 1]
```diff
@@ -1,23 +1,12 @@
...
\ No newline at end of file
+namespace Microsoft.AspNetCore.Razor.Runtime.TagHelpers
+{
+    public partial class TagHelperExecutionContext
+    {
+        internal TagHelperExecutionContext(string tagName, Microsoft.AspNetCore.Razor.TagHelpers.TagMode tagMode) { }
+        [System.Diagnostics.DebuggerStepThroughAttribute]
+        internal System.Threading.Tasks.Task<Microsoft.AspNetCore.Razor.TagHelpers.TagHelperContent> GetChildContentAsync(bool useCachedResult, System.Text.Encodings.Web.HtmlEncoder encoder) { throw null; }
+    }
+}
```
Comment (halter73) [inline src/Razor/Razor.Runtime/ref/Microsoft.AspNetCore.Razor.Runtime.Manual.cs]: @dougbu So far this is the only changed (as apposed to added or deleted) Manual.cs file in the merge. It seems like this change was made in release/3.1. Is this expected?

@JunTaoLuo

**Comment by ajaybhargavb** [comment]
Current build failures in Kestrel,
```
##[error]Http3/Http3TestBase.cs(343,102): error CS1929: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Span<byte>' does not contain a definition for 'GetAsciiOrUTF8StringNonNullCharacters' and the best extension method overload 'HttpUtilities.GetAsciiOrUTF8StringNonNullCharacters(ReadOnlySpan<byte>)' requires a receiver of type 'ReadOnlySpan<byte>'
```
```
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(474,71): error CS0234: The type or namespace name 'IHttpParser<>' does not exist in the namespace 'Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http' (are you missing an assembly reference?) [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(474,71): error CS0234: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'IHttpParser<>' does not exist in the namespace 'Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http' (are you missing an assembly reference?)
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnHeader(ReadOnlySpan<byte>, ReadOnlySpan<byte>)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnHeader(ReadOnlySpan<byte>, ReadOnlySpan<byte>)'
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnHeadersComplete(bool)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnHeadersComplete(bool)'
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int)'
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int, ReadOnlySpan<byte>)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1216,46): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http2Connection' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int, ReadOnlySpan<byte>)'
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnHeader(ReadOnlySpan<byte>, ReadOnlySpan<byte>)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnHeader(ReadOnlySpan<byte>, ReadOnlySpan<byte>)'
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnHeadersComplete(bool)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnHeadersComplete(bool)'
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int)'
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int, ReadOnlySpan<byte>)' [/home/vsts/work/1/s/src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
##[error]Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1105,60): error CS0535: (NETCORE_ENGINEERING_TELEMETRY=Build) 'Http1ParsingHandler' does not implement interface member 'IHttpHeadersHandler.OnStaticIndexedHeader(int, ReadOnlySpan<byte>)'
```
@halter73 @jkotalik, can you take a look?

Failure in build,
```
##[error]src/Framework/src/Microsoft.AspNetCore.App.Runtime.csproj(387,7): error MSB4064: (NETCORE_ENGINEERING_TELEMETRY=Build) The "Uri" parameter is not supported by the "DownloadFile" task. Verify the parameter exists on the task, and it is a settable public instance property.
```

@dougbu @wtgodbe, looks like the custom `DownloadFile` task was indeed being used in App.Runtime.csproj. Should we bring it back?

**Comment by ajaybhargavb** [comment]
Also, the CI source build failed with,
```
##[error]eng/targets/ResolveReferences.targets(208,5): error MSB3245: (NETCORE_ENGINEERING_TELEMETRY=Restore) Could not resolve this reference. Could not locate the package or project for "Microsoft.Bcl.AsyncInterfaces". Did you update baselines and dependencies lists? See docs/ReferenceResolution.md for more details.
```
cc @JunTaoLuo

**Comment by dougbu** [comment]
> @dougbu @wtgodbe, looks like the custom DownloadFile task was indeed being used in App.Runtime.csproj. Should we bring it back?

Yes, we goofed. Added this task just recently in 'release/3.1' and need it in 'master'.

File: src/Razor/Razor.Runtime/ref/Microsoft.AspNetCore.Razor.Runtime.Manual.cs (hunk 1 above)
Comment (dougbu) [reply src/Razor/Razor.Runtime/ref/Microsoft.AspNetCore.Razor.Runtime.Manual.cs]: Best to undo all changes to *.Manual.cs files then regenerate the ref/ projects (again). We don't need the `internal` APIs in 'master' and including them will make everything much more brittle.

Comment (ajaybhargavb) [reply src/Razor/Razor.Runtime/ref/Microsoft.AspNetCore.Razor.Runtime.Manual.cs]: I reverted this file to master and re-ran GenerateReferenceAssemblies.ps1.

**Comment by ajaybhargavb** [comment]
@dougbu I added back DownloadFile task.

@halter73 @jkotalik, I'll let you handle the kestrel build failures here.

**Comment by BrennanConroy** [comment]
Conflicting file: .azure/pipelines/ci.yml

**Comment by ajaybhargavb** [comment]
@JunTaoLuo and I tried to resolve this conflict but we're not sure what `group` should resolve to in `ci.yml`. @HaoK @wtgodbe can you guys give this a shot?

**Comment by HaoK** [comment]
Okay I resolved it I think

**Comment by halter73** [comment]
The Kestrel failures are due to internal changes relative to the Kestrel.Core.Manual.cs ref file that was added to release/3.1. I think these Manual.cs files should only be checked into servicing branches, since I don't think we're concerned about API changes to internal types/methods/properties in master. I'm not sure though, since there seem to be Manual.cs files already checked into master.

**Review by halter73 — APPROVED** [review]

File: .azure/pipelines/jobs/default-build.yml [hunk 2]
```diff
@@ -161,6 +161,24 @@ jobs:
...
+              env:
+                Token: $(dn-bot-dnceng-artifact-feeds-rw)
+        - ${{ if ne(parameters.agentOs, 'Windows') }}:
+          - task: Bash@3
+            displayName: Setup Private Feeds Credentials
+            inputs:
+              filePath: $(Build.SourcesDirectory)/eng/common/SetupNugetSources.sh
+              arguments: $(Build.SourcesDirectory)/NuGet.config $Token
+            env:
+              Token: $(dn-bot-dnceng-artifact-feeds-rw)
```
Comment (dougbu) [inline .azure/pipelines/jobs/default-build.yml]: Why does this block need to be duplicated in ci.yml for anything but the source build job? (That one doesn't use this template.)

File: .azure/pipelines/ci.yml [hunk 3]
```diff
@@ -641,6 +699,8 @@ stages:
 
   # Source build
   - job: Source_Build
+    # Skipping until someone can look into this
+    condition: false
```
Comment (dougbu) [inline .azure/pipelines/ci.yml]: I'll look at this later tonight

File: src/Azure/AzureAppServicesIntegration/src/Microsoft.AspNetCore.AzureAppServicesIntegration.csproj [hunk 4]
```diff
@@ -7,7 +7,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <PackageTags>aspnetcore;azure;appservices</PackageTags>
-    <IsShippingPackage>true</IsShippingPackage>
+    <IsPackable>true</IsPackable>
```
Comment (dougbu) [inline src/Azure/AzureAppServicesIntegration/src/Microsoft.AspNetCore.AzureAppServicesIntegration.csproj]: Why do a `$(IsShippingPackage)` -> `$(IsPackable)` switch in projects that aren't marked as `$(IsAspNetCoreApp)`? This project shouldn't have a ref/ project associated with it.

File: src/Components/Analyzers/src/Microsoft.AspNetCore.Components.Analyzers.csproj [hunk 5]
```diff
@@ -6,7 +6,7 @@
     <NoPackageAnalysis>true</NoPackageAnalysis>
     <GenerateDocumentationFile>false</GenerateDocumentationFile>
     <Description>Roslyn analyzers for ASP.NET Core Components.</Description>
-    <IsShippingPackage>true</IsShippingPackage>
+    <IsPackable>true</IsPackable>
```
Comment (dougbu) [inline src/Components/Analyzers/src/Microsoft.AspNetCore.Components.Analyzers.csproj]: Huh? Don't we want to ship this analyzer?

File: src/DataProtection/AzureKeyVault/src/Microsoft.AspNetCore.DataProtection.AzureKeyVault.csproj [hunk 6]
```diff
@@ -5,7 +5,7 @@
     <TargetFramework>netstandard2.0</TargetFramework>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <PackageTags>aspnetcore;dataprotection;azure;keyvault</PackageTags>
-    <IsShippingPackage>true</IsShippingPackage>
+    <IsPackable>true</IsPackable>
```
Comment (dougbu) [inline src/DataProtection/AzureKeyVault/src/Microsoft.AspNetCore.DataProtection.AzureKeyVault.csproj]: Bet we want to ship this project too

File: src/Components/Analyzers/src/Microsoft.AspNetCore.Components.Analyzers.csproj (hunk 5 above)
Comment (dougbu) [reply src/Components/Analyzers/src/Microsoft.AspNetCore.Components.Analyzers.csproj]: Ooops, s/ship/pack/

Comment (dougbu) [reply src/Components/Analyzers/src/Microsoft.AspNetCore.Components.Analyzers.csproj]: Nevermind. Both of these setting match the defaults and could be left out entirely.

**Comment by dougbu** [comment]
FYI My bet is a lot of the *.csproj changes in 2d066dcd3e32, "Apply some fixes" could be reverted. The `$(IsShippingPackage)` -> `$(IsPackable)` changes shouldn't matter because both settings are to the default (`true)`. Suggest we do a follow up to remove `<IsShippingPackage>true</IsShippingPackage>` from all src/*.csproj files.

**Comment by dougbu** [comment]
Haven't figured out the overall template test failures yet. Suspect Microsoft.AspNetCore.App or Microsoft.AspNetCore.App.Ref packages are incomplete or otherwise messed up. Will download them and look tomorrow morning.

**Comment by dougbu** [comment]
One templates problem _may_ be related to lack of System.* ref/ assemblies in Microsoft.AspNetCore.App.Ref (Microsoft.Win32.SystemEvents.dll is missing too). @JunTaoLuo has a PR (#19161) out fixing this in 'release/3.1' and we might want to cherry-pick that into this branch once it's merged.

In any case, #19161 will be useful in the 'master' branch but might not fix the template tests.

**Review by dougbu — APPROVED** [review]

**Review by dougbu — APPROVED** [review]

**Comment by dougbu** [comment]
TODO items for after this is merged:
- [ ] file issue about removing `<IsPackable>true</IsPackable>` and similar "set to the default" property settings in implementation projects -- or do the needful
- [ ] file issue about removing duplicate "Setup Private Feeds Credentials" build steps between ci.yml and default-build.yml -- or do the needful

@ajaybhargavb @BrennanConroy @halter73 @JunTaoLuo @wtgodbe and anyone else who knows of a follow-up action I missed, please edit this comment

**Comment by dougbu** [comment]
🎉 🎉 🎉

**Pull request merged**